test/contrib/automotive/xcp/xcp.uts

```diff
@@ -0,0 +1,823 @@
+assert cto_response.packet_code == 0xFF
+assert cto_response.answers(cto_request)
+
+program_start_response = cto_response['ProgramStartPositiveResponse']
+
+assert program_start_response.comm_mode_pgm == 0x01
+assert "master_block_mode" in program_start_response.comm_mode_pgm
+assert "interleaved_mode" not in program_start_response.comm_mode_pgm
+assert "x2" not in program_start_response.comm_mode_pgm
+assert "x3" not in program_start_response.comm_mode_pgm
+assert "x4" not in program_start_response.comm_mode_pgm
+assert "x5" not in program_start_response.comm_mode_pgm
+assert "slave_block_mode" not in program_start_response.comm_mode_pgm
+assert "x7" not in program_start_response.comm_mode_pgm
+
+assert program_start_response.max_cto_pgm == 0x08
+assert program_start_response.max_bs_pgm == 0x2a
+assert program_start_response.min_bs_pgm == 0xff
+assert program_start_response.queue_size_pgm == 0xdd
+
+= test_program_clear(self):
+conf.contribs['XCP']['byte_order'] = 0
+
+cto_request = CTORequest(pid="PROGRAM_CLEAR") / ProgramClear(b'\x00\xff\xff\x00\x01\x00\x00')
+assert cto_request.pid == 0xd1
+assert bytes(cto_request) == b'\xd1\x00\xff\xff\x00\x01\x00\x00'
+
+assert cto_request['ProgramClear'].mode == 0x00
+assert cto_request['ProgramClear'].clear_range == 0x00000100
+
+cto_response = CTOResponse(b'\xff')
+assert cto_response.packet_code == 0xFF
+assert cto_response.answers(cto_request)
+
+= test_program
+
+conf.contribs['XCP']['byte_order'] = 0
+conf.contribs['XCP']['MAX_CTO'] = 8
+conf.contribs['XCP']['Address_Granularity_Byte'] = 1
+
+cto_request = CTORequest(pid="PROGRAM") / Program(b'\x06\x00\x01\x02\x03\x04\x05')
+assert cto_request.pid == 0xd0
+assert bytes(cto_request) == b'\xd0\x06\x00\x01\x02\x03\x04\x05'
+
+assert cto_request['Program'].nr_data_element == 0x06
+assert cto_request['Program'].data_elements == b"\x00\x01\x02\x03\x04\x05"
+
+cto_response = CTOResponse(b'\xff')
+assert cto_response.packet_code == 0xFF
+assert cto_response.answers(cto_request)
+
+
++ Tests on a virtual CAN-Bus
+= Imports
+
+import threading
+import time
+from subprocess import call
+
+import six
+
+from scapy.consts import LINUX
+from scapy.contrib.automotive.xcp.cto_commands_master import Connect, TransportLayerCmd, TransportLayerCmdGetSlaveId
+from scapy.contrib.automotive.xcp.scanner import XCPOnCANScanner
+from scapy.contrib.automotive.xcp.xcp import CTORequest, XCPOnCAN, CTOResponse, ConnectPositiveResponse, TransportLayerCmdGetSlaveIdResponse, GenericResponse
+from scapy.contrib.cansocket_python_can import CANSocket
+from scapy.main import load_layer
+
+= Load module
+
+load_layer("can")
+load_layer("automotive.xcp.xcp",  globals_dict=globals())
+
+= Global variables
+
+iface0 = "vcan0"
+
+= Initialize a virtual CAN interface
+
+def set_up_virtual_python_can_socket(iface):
+    return CANSocket(bustype='virtual', channel=iface, timeout=0.01)
+
+print('setting up CAN')
+
+if 0 != call("cansend %s 000#" % iface0, shell=True):
```

Comment:
Please use this style of using `call` to avoid `shell=True`.
Probably it's a good idea to search for `shell=True` in all XCP related files and change them.

Reply:
What do you mean? Which style of using call should I use?

Reply:
Wups, sorry, I forgot to link it.
For example here
https://github.com/secdev/scapy/blob/master/test/contrib/automotive/obd/scanner.uts#L30

I once made a bit PR where I refactored all automotive related tests:
https://github.com/secdev/scapy/commit/d3fc881dae1a0d4c603090b9bb29432579f33688
There I removed all `shell=True` calls and also `sys.exit(-1)`